src/include/grpc_transcoding/message_reader.h

```diff
@@ -70,7 +74,20 @@ class MessageReader {
   //       of the original ZeroCopyInputStream and the MessageReader relies on
   //       the caller to advance the stream to the next message before calling
   //       NextMessage() again.
-  std::unique_ptr<::google::protobuf::io::ZeroCopyInputStream> NextMessage();
+  // NOTE: the caller should check `Status()` is OK after calling this method.
+  std::unique_ptr<::google::protobuf::io::ZeroCopyInputStream> NextMessage() {
+    // Buffer is not used.
+    unsigned char* delimiter_buffer;
+    return NextMessage(&delimiter_buffer);
+  }
+
+  // An overload that also outputs the gRPC message delimiter for the parsed
+  // message. The caller must NOT take ownership of the buffer. The buffer is
+  // guaranteed to have `kGrpcDelimiterByteSize` bytes.
+  // NOTE: the caller must check the return is NOT nullptr before consuming
+  //       the `delimiter_buffer`.
+  std::unique_ptr<::google::protobuf::io::ZeroCopyInputStream> NextMessage(
+      unsigned char** delimiter_buffer);
```

Comment:
Just point the buffer from internal implementation for callers to use is dangerous.  the ownership of the buffer is not clear.  the buffer data can be overrided later.  

Just 5 byte data, we can define a struct to copy it out as unique_ptr. 

can we return 

std::pair<ZeroCopyInput,  GrpcDelimitor>

Reply:
Agreed this is dangerous, but I'm worried about making so many copies, even if it's 5 bytes. This pattern of exposing internal buffers is used throughout this repo and the proto repo.

In envoy, we will make a copy anyways of the `delimiter_buffer` via `getBuffer().add(...)`. So returning it would result in 2 `memcpy` of the 5 bytes.

```
unsigned char* delimiter_buffer;
  std::unique_ptr<ZeroCopyInputStream> message =
      request_reader_.NextMessage(&delimiter_buffer);

  RETURN_IF_ERROR(request_reader_.Status());
  if (message == nullptr) {
    return absl::OkStatus();
  }

  request_out_.getBuffer().add(
      delimiter_buffer, google::grpc::transcoding::kGrpcDelimiterByteSize);
```

What do you think? If you're ok with 2 copies, that's fine by me.

Reply:
number of memcpy is not my concern since it is only 5 bytes.

My concern is function interface, or API,   is it clean,  is it error prone? 

with current way, it is passing pointer around, and size is defined separately

My idea is:  define a class

class  GrpcFrameHeader {
   public:
      const char * datea() const;
      size_t size() const;
   private:
      char data[kGrpcDelimiterByteSize];
};

Or use ZeroCopyInputStream for the 5 byte data.   But it seems that it is overkill.


Then I prefer the "NextMessage" function to return two unique_ptr, one for the message, another is GrpcFrameHeader.  But c++ function can only return one object.  One for return, another one in the funciton argument?  Well, I don't like it.

Ok,  how about this

std::unique_ptr<ZeroCopyInputStream> NextMessage(std::unique_ptr<GrpcFrameHeader>& grpc_frame);

Reply:
Ok, I actually prefer your previous solution of returning a Struct. Having a reference to a `unique_ptr` seems like an antipattern.

I returned a fixed-size array in the struct. Now the size information is provided at compile time. Callers can just use `sizeof(grpc_frame)` to read the correct size. Using `ZeroCopyInputStream` is not needed for storing the size.

Reply:
Yes, I like this too. Thanks